File: python/longest-palindrome.py
class Solution:
    """
    @param s: a string which consists of lowercase or uppercase letters
    @return: the length of the longest palindromes that can be built
    """

    def longestPalindrome(self, s):
        # write your code here

        # find how many pair
        findPairSet = set()
        pairs = 0

        for char in s:
            if char in findPairSet:
                findPairSet.remove(char)
                pairs += 1
            else:
                findPairSet.add(char)

        return pairs * 2 + 1 if len(findPairSet) > 0 else pairs * 2


class Solution:
    """
    @param s: a string which consists of lowercase or uppercase letters
    @return: the length of the longest palindromes that can be built
    """

    def longestPalindrome(self, s):
        # write your code here
        hashmap = {}

        for l in s:
            if l not in hashmap:
                hashmap[l] = 1
            else:
                hashmap[l] += 1

        hasSingle = False
        res = 0
        for l in hashmap.keys():
            if not hasSingle and (hashmap[l] == 1 or hashmap[l] % 2 == 1):
                hasSingle = True
            res += hashmap[l] // 2 * 2

        return res + 1 if hasSingle else res

File: java/LongestPalindrome.java
public class Solution {
    // https://www.lintcode.com/problem/longest-palindrome/description
    /**
     * @param s: a string which consists of lowercase or uppercase letters
     * @return: the length of the longest palindromes that can be built
     */
    public int longestPalindrome(String s) {
        // write your code here

        HashSet<Character> set = new HashSet<>();
        
        for(char c: s.toCharArray()) {
            if(set.contains(c)) {
                set.remove(c);
            } else {
                set.add(c);
            }
        }
        
        int singleChars = set.size();
        
        if(singleChars > 0) {
            singleChars--;
        }
        
        return s.length() - singleChars;

        // HashMap<Character, Integer> map = new HashMap<>();
        
        // for(char c: s.toCharArray()) {
        //     map.put(c, map.getOrDefault(c, 0) + 1);
        // }

        // int pair = 0;
        // int single = 0;
        
        // for(int count : map.values()) {
        //     int div = count / 2;
        //     pair += div;
        //     single += count % 2;
        // }
        
        // return (pair * 2) + (single > 0 ? 1 : 0);
    }
}